### ISSUE TITLE
Add feature to cause nutrition decay to result in fatness increase.

### ISSUE THREAD

**Issue by SpadesNeil** [SpadesNeil/VOREstation#357]
While nutrition is above 500, the player is considered to have been over-eating, and so as nutrition decays, it causes fatness to tick upward at a weightGain rate specified by the player between 0% (no weight gain) to 100% (realistic weight gain). This will be part of player options in the future.

Note the following:
- 1 same size person is worth about 1070 nutriment.
- IRL the human body has about 110,000 calories.
- IRL 3,500 calories = 1 pound, or 7,700 calories = 1 kilogram.
- Therefore, 110,000 / 1070 = ~103.
- Thus, for every 103 nutriment absorbed while overfed, the character gains the equivalent of 31 pounds.
- 103/31=3.3. 1 nutriment = 3.3 pounds. Therefore each nutriment absorbed while overweight does the following math: (Nutriment * 3.3) * weightGain = addedWeight.
- While nutriment decays, if overweight, add 3.3 pounds for every 1 nutriment absorbed assuming the weight gain is set to 1.0 (100%) which it will be by default.

Range between 130-140 pounds is chosen as the default weight (based on 137 average for a human body).

Maybe considering a whole person is worth ~1070 nutriment, we could add an additional equation to determine how much nutriment is gained depending on a subject's actual weight. If 1070 nutriment = 136 pounds, then perhaps some additional math is in order to decide how much weight is gained from digesting someone. Every pound of the victim is ultimately worth 7.8, which if we leave the code unchanged, means ~1 nutriment per 8 pounds of the eaten character's weight. So someone who eats (for example) Rosey, is going to gain more weight than if they ate Greyson. The nutriment gain could be altered through this equation. Assuming we use pounds:

(137 - prey.weight) / 137 = The percentage offset between default weight and actual weight. So if someone is 10% fatter than the average prey, they'll yield 10% more nutriment per digestion tick. There's a few ways the math for this could work out.

if(percentOffset)
   nutriment += 1*percentOffset
else // no offset)
   nutriment += 1

Conversely, there will also be a weight loss mechanics. If nutriment hits 0, every burn of nutriment loss will instead be drawn from your fat reserves.

Future features also to be added later include weight loss equipment which will have minigames to keep it both interesting and slightly tedious. There will also be a scale in medbay to directly calculate weight.

The character's current weight will be reflected via examine text just like nutriment levels are for stomach fullness.

Character base weight CANNOT go below 70 pounds or above 500 pounds, because after either end of those extremes, you hit some serious problems that we can't simulate in-game, and they go to extreme fringe of fetishes that push the boundaries of peoples' tolerances to the grotesque.

I might change the weight to match with the multiplications done by the size changing mechanics, so a 100lb person only weighs 25lbs while shrunken down to the minimum size which is 0.25 of base size.

**This feature is part of a bigger Body Weight Mechanic update and is not to be considered a complete feature with this issue alone. See the milestone for more information.**

**Comment by SpadesNeil** [comment]
Changed to blocker because the feature can't exist without this.

**Comment by SpadesNeil** [comment]
CTRL+F to find snips that need work in the branch.
>VOREstation weight code TODO.

**Comment by SpadesNeil** [comment]
Okay I forgot some math. 1070 = 31 pounds, so 1 nutrient = ~0.3 pounds.

**Comment by SpadesNeil** [comment]
@Cameron653 and @antsnap want to help me out here? I think this part is done.

**Comment by Cameron653** [comment]
I'm really, uh, inexperienced with this, but, looking through it, it seems like it'd all work. 
I'm not 100% sure, but from what I'm seeing would work.

I ran some numbers through it to make sure it'd work correctly, too.

Example: 100 starting pounds, metabolism set to 1, starting nutrition is 400
offset = (137 - M.weight) / 137
offset = (137 - 100) / 137
offset = 37/137
offset = .27

var/difference = owner.playerscale / M.playerscale
var/difference = 1 / 2 //Normal size eating a macro, aka big guy.
owner.nutrition += offset*(10/difference)
owner.nutrition += .27*(10/.5)
owner.nutrition += 5.4 //They gain 5.4 nutrition per digestion tick.

if (nutrition > 0 && stat != 2)
nutrition = max (0, 400 - HUNGER_FACTOR)
nutrition = max (0, 400 - 0.05) //I think I'm doing this right?

weight = metabolism*(0.3*(max (0, nutrition - HUNGER_FACTOR))) //Giving them 400 nutrition+above
weight = 1*(.3*(max (0, 405.4 - 0.05))) // going to ignore the zero
weight = 1*(.3*(399.95 (405.4 - 0.05)))
weight = 1*(.3 (399.95) //They have too high of a nutrition, so it goes to the max.
weight = 1*(119.985) //They gained 19.985 pounds from 5.4 nutrition. 3.7 per unit.
weight = 119.985

It all looks like it checks out.

**Comment by SpadesNeil** [comment]
>They're getting 5.4 nutrition per digestion tick. Not digestion. Meaning you could easily hit the 500 pound cap. You might want to make it lower.

~20-30 pound weight gain is right where I want it to be per victim for 100% realistic gain. Players can adjust it.

**Comment by Cameron653** [comment]
> ~20-30 pound weight gain per victim is right where I want it to be per victim for 100% realistic gain. Players can adjust it. Are you referring to something else?

Lines 156-167 in belly_stomach.dm. It's giving them 5.4 nutrition every time digestion **_damage_** is dealt. Meaning that they gain 16 pounds per tick (Assuming the prey is 2x bigger than the pred). If it was doing 5 damage per digestion tick (Not accounting for natural healing), they would gain 320 pounds from 1 person if they immediately died as soon as they hit -0 hp. 640 pounds if they died when they hit -100 hp (Still not accounting for natural healing)

lines 158-167 should probably be moved upwards to when someone digests, not when they take damage. At the moment, the pred gets a flat out 20 nutrition for digesting a person/mob/whatnot, as shown by line 81 in belly_stomach. Not accounting for any digestion damage they dealt to the prey before that.


Basically,

https://github.com/SpadesNeil/VOREstation/blob/5b2b14a062930c4151b49ea16e4de4f0620ba906/code/modules/vore/belly/belly_stomach.dm#L156-L167

should be moved to 

https://github.com/SpadesNeil/VOREstation/blob/5b2b14a062930c4151b49ea16e4de4f0620ba906/code/modules/vore/belly/belly_stomach.dm#L81

**Comment by SpadesNeil** [comment]
See this is why I never claim to be a good coder. Just a persistent one.

Let's backtrack because I probably screwed this up. A fully digested victim renders 1070 nutriment. The goal is to make 1070 worth 30 pounds of weight gain as 1070 decays away (assuming the predator has already eaten enough to be "full" and gain all the weight).

I clearly fucked up the math because 0.3 * 1070 = 321, so that's not right.

What if I did 0.03?

**Comment by Cameron653** [comment]
Example: 137 starting pounds, metabolism set to 1, starting nutrition is 400
offset = (137 - 137) / 137
offset = (137 - 137) / 137
offset = 0/137
offset = 0

var/difference = owner.playerscale / M.playerscale
var/difference = 1 / 1 //Normal size eating a normal sized person, and they have exactly 137 pounds
owner.nutrition += (26.75/difference)
owner.nutrition += (26.75/1)
owner.nutrition += 26.75 //They gain 26.75 nutrition per digestion tick. Person dies after 40 digestion ticks (Assuming no natural healing), meaning they get 1070 nutrition

if (nutrition > 0 && stat != 2)
nutrition = max (0, 1470 - HUNGER_FACTOR)
nutrition = max (0, 1470 - 0.05) //I think I'm doing this right?

weight += metabolism(0.03(max (0, nutrition - HUNGER_FACTOR))) 
weight += 1 * (.03 *  (1470 - 0.05)) // going to ignore the zero
weight += 1 * (.03 *  (1470 - 0.05))
weight += 1 * (.03 * (1476.95)
weight += 1 * (32.1) //They gained 19.985 pounds from 5.4 nutrition. 3.7 per unit.
weight += 32.1

**Comment by SpadesNeil** [comment]
I have come to the firm conclusion I am fucking terrible at math but I may have found a solution.

**Comment by SpadesNeil** [comment]
OKAY FINALLY THIS BULLSHIT IS RESOLVED.

**Comment by SpadesNeil** [comment]
THERE. FUCK YOU, _NUMBERS_. YOU AND YOUR **MATH**.

Okay moving on. Now that the framework is laid, we can do the rest. Yay!
